### PR TITLE
Support --header option

### PR DESCRIPTION
**Pull request by darnir**
* src/options.c: Add new --header option and a new parse_header()
parsing function for that
* src/wget_options.h: Same
* src/wget.c (http_create_request): Add the user added headers to the
HTTP Request
* src/test-wget-1.c: Add a test for the --header option

### CODE REVIEW

**Comment by darnir** [comment]
The test for this currently simply sends the Range header to the server and expects a partial response.

A more thorough test would involve adding a new option, "WGET_TEST_EXPECTED_HEADERS", which contains an array of strings of expected headers and the server must verify that each of the expected headers has arrived. The server must also be able to test that no extra headers have arrived, so it must be preloaded with a set of known, expected headers.

**Comment by rockdaboot** [comment]
Good work.
Valgrind still complains:
==26188== Invalid read of size 4
==26188==    at 0x4E5FD85: wget_vector_size (vector.c:340)
==26188==    by 0x40A5C7: selftest_options (options.c:1805)
==26188==    by 0x4054D3: main (test.c:1803)
==26188==  Address 0xa37b80c is 28 bytes inside a block of size 40 free'd
==26188==    at 0x4C2BDDB: free (vg_replace_malloc.c:530)
==26188==    by 0x4E5FD18: wget_vector_free (vector.c:302)
==26188==    by 0x407E49: parse_header (options.c:364)
==26188==    by 0x40875C: set_long_option (options.c:899)
==26188==    by 0x408A4D: parse_command_line (options.c:1139)
==26188==    by 0x40A5BB: selftest_options (options.c:1804)
==26188==    by 0x4054D3: main (test.c:1803)
==26188==  Block was alloc'd at
==26188==    at 0x4C2CBC5: calloc (vg_replace_malloc.c:711)
==26188==    by 0x4E601A8: wget_calloc (xalloc.c:108)
==26188==    by 0x4E5F870: wget_vector_create (vector.c:64)
==26188==    by 0x407E9B: parse_header (options.c:335)
==26188==    by 0x40875C: set_long_option (options.c:899)
==26188==    by 0x408A4D: parse_command_line (options.c:1139)
==26188==    by 0x40A4E2: selftest_options (options.c:1773)
==26188==    by 0x4054D3: main (test.c:1803)
==26188== 
==26188== Invalid read of size 4
==26188==    at 0x4E5FD85: wget_vector_size (vector.c:340)
==26188==    by 0x40A260: deinit (options.c:1594)
==26188==    by 0x4054E7: main (test.c:1805)
==26188==  Address 0xa37b80c is 28 bytes inside a block of size 40 free'd
==26188==    at 0x4C2BDDB: free (vg_replace_malloc.c:530)
==26188==    by 0x4E5FD18: wget_vector_free (vector.c:302)
==26188==    by 0x407E49: parse_header (options.c:364)
==26188==    by 0x40875C: set_long_option (options.c:899)
==26188==    by 0x408A4D: parse_command_line (options.c:1139)
==26188==    by 0x40A5BB: selftest_options (options.c:1804)
==26188==    by 0x4054D3: main (test.c:1803)
==26188==  Block was alloc'd at
==26188==    at 0x4C2CBC5: calloc (vg_replace_malloc.c:711)
==26188==    by 0x4E601A8: wget_calloc (xalloc.c:108)
==26188==    by 0x4E5F870: wget_vector_create (vector.c:64)
==26188==    by 0x407E9B: parse_header (options.c:335)
==26188==    by 0x40875C: set_long_option (options.c:899)
==26188==    by 0x408A4D: parse_command_line (options.c:1139)
==26188==    by 0x40A4E2: selftest_options (options.c:1773)
==26188==    by 0x4054D3: main (test.c:1803)
==26188== 
==26188== Invalid read of size 8
==26188==    at 0x4E5FCE1: wget_vector_free (vector.c:298)
==26188==    by 0x40A26E: deinit (options.c:1600)
==26188==    by 0x4054E7: main (test.c:1805)
==26188==  Address 0xa37b800 is 16 bytes inside a block of size 40 free'd
==26188==    at 0x4C2BDDB: free (vg_replace_malloc.c:530)
==26188==    by 0x4E5FD18: wget_vector_free (vector.c:302)
==26188==    by 0x407E49: parse_header (options.c:364)
==26188==    by 0x40875C: set_long_option (options.c:899)
==26188==    by 0x408A4D: parse_command_line (options.c:1139)
==26188==    by 0x40A5BB: selftest_options (options.c:1804)
==26188==    by 0x4054D3: main (test.c:1803)
==26188==  Block was alloc'd at
==26188==    at 0x4C2CBC5: calloc (vg_replace_malloc.c:711)
==26188==    by 0x4E601A8: wget_calloc (xalloc.c:108)
==26188==    by 0x4E5F870: wget_vector_create (vector.c:64)
==26188==    by 0x407E9B: parse_header (options.c:335)
==26188==    by 0x40875C: set_long_option (options.c:899)
==26188==    by 0x408A4D: parse_command_line (options.c:1139)
==26188==    by 0x40A4E2: selftest_options (options.c:1773)
==26188==    by 0x4054D3: main (test.c:1803)
==26188== 
==26188== Invalid free() / delete / delete[] / realloc()
==26188==    at 0x4C2BDDB: free (vg_replace_malloc.c:530)
==26188==    by 0x4E5FD18: wget_vector_free (vector.c:302)
==26188==    by 0x40A26E: deinit (options.c:1600)
==26188==    by 0x4054E7: main (test.c:1805)
==26188==  Address 0xa37b7f0 is 0 bytes inside a block of size 40 free'd
==26188==    at 0x4C2BDDB: free (vg_replace_malloc.c:530)
==26188==    by 0x4E5FD18: wget_vector_free (vector.c:302)
==26188==    by 0x407E49: parse_header (options.c:364)
==26188==    by 0x40875C: set_long_option (options.c:899)
==26188==    by 0x408A4D: parse_command_line (options.c:1139)
==26188==    by 0x40A5BB: selftest_options (options.c:1804)
==26188==    by 0x4054D3: main (test.c:1803)
==26188==  Block was alloc'd at
==26188==    at 0x4C2CBC5: calloc (vg_replace_malloc.c:711)
==26188==    by 0x4E601A8: wget_calloc (xalloc.c:108)
==26188==    by 0x4E5F870: wget_vector_create (vector.c:64)
==26188==    by 0x407E9B: parse_header (options.c:335)
==26188==    by 0x40875C: set_long_option (options.c:899)
==26188==    by 0x408A4D: parse_command_line (options.c:1139)
==26188==    by 0x40A4E2: selftest_options (options.c:1773)
==26188==    by 0x4054D3: main (test.c:1803)

**Comment by darnir** [comment]
Aah.. I wrote that last minute and didn't check it with valgrind.

I'll fix it later today

Thanking You,
Darshit Shah

**Comment by rockdaboot** [comment]
Sorry, found something we should take care of... from the wget2 manual (docs/wget2_manual.md):
```
  Specification of an empty string as the header value will clear all previous user-defined headers.

  As of Wget 1.10, this option can be used to override headers otherwise generated automatically.  This example
  instructs Wget to connect to localhost, but to specify foo.bar in the "Host" header:

      wget --header="Host: foo.bar" http://localhost/
```
Can you take care of that ?

BTW, that reminds me to go through wget2_manual.md before the release.

File: src/options.c [hunk 1]
```diff
@@ -314,6 +315,58 @@ static int parse_stringset(option_t opt, const char *val)
...
+		while (*value != '\0' && *value == ' ')
+			value++;
+
+		if (*value == '\0') {
+			wget_error_printf("No value in header (ignoring): %s\n", val);
+			return 0;
+		}
+
+		_param.name = wget_strmemdup(val, delim_pos - val);
+		_param.value = strdup(value);
```
Comment (rockdaboot) [inline src/options.c]: Please use wget_strdup, for OOM checking.

File: src/options.c [hunk 2]
```diff
@@ -314,6 +315,58 @@ static int parse_stringset(option_t opt, const char *val)
...
+
+		delim_pos = strchr(val, ':');
+
+		if (!delim_pos || delim_pos == val) {
+			wget_error_printf("Ignoring invalid header: %s\n", val);
+			return 0;
+		}
+
+		value = delim_pos + 1;
+		while (*value != '\0' && *value == ' ')
```
Comment (rockdaboot) [inline src/options.c]: while (*value == ' ') should be enough.

File: src/options.c [hunk 3]
```diff
@@ -1537,6 +1591,14 @@ void deinit(void)
 	wget_vector_free(&config.accept_patterns);
 	wget_vector_free(&config.reject_patterns);
 
+	for (int i = 0; i < wget_vector_size(config.headers); i++) {
+		wget_http_header_param_t *p = wget_vector_get(config.headers, i);
+		xfree(p->name);
+		xfree(p->value);
+	}
+
```
Comment (darnir) [inline src/options.c]: I was already planning on doing that. I'd mentioned it in the TODO in the last commit message. 

Anyhow, I've pushed a new commit with all the above changes. It's ready for merge if you're okay

**Pull request merged**